source_code/main_mcu/src/TIMER/driver_timer.c

```diff
@@ -221,6 +221,49 @@ void timer_set_calendar(uint16_t year, uint16_t month, uint16_t day, uint16_t ho
 #endif
 }
 
+/*! \fn	time(void)
+*   \brief  Get the current unix time
+*   \param  None
+*   \return Seconds since 1970
+*/
+uint64_t time(void)
+{
+    uint64_t unix_time = 0;
+    int32_t i;
+    uint32_t num_days = 0;
+    calendar_t tm;
+    uint8_t is_curr_year_leap;
+    uint16_t year;
+
+    timer_get_calendar(&tm);
+    year = tm.bit.YEAR + 2000;
+    is_curr_year_leap = IS_LEAP_YEAR(year);
+
+    //Add up days in all years
+    for (i = EPOCH_YEAR; i < year; ++i)
+    {
+        num_days += IS_LEAP_YEAR(i) ? 366 : 365;
+    }
+
+    //Add up days in all months
+    for (i = 0; i < tm.bit.MONTH - 1; ++i)
+    {
+        num_days += dph[i];
+        num_days += ((i == LEAP_MONTH) && is_curr_year_leap) ? 1 : 0;
+    }
+
+    //Add days in month
+    num_days += tm.bit.DAY - 1;
+
+    //Add the days hours, minutes, and seconds
+    unix_time = num_days * SEC_IN_DAY;
+    unix_time += tm.bit.HOUR * SEC_IN_HOUR;
+    unix_time += tm.bit.MINUTE * 60;
+    unix_time += tm.bit.SECOND;
+
+    return unix_time;
```

Comment:
but... what about leap seconds? :p

Reply:
Good question. See [1]. 

"Coordinated Universal Time (UTC) includes leap seconds. However, in POSIX time (seconds since the Epoch), leap seconds are ignored (not applied) to provide an easy and compatible method of computing time differences. Broken-down POSIX time is therefore not necessarily UTC, despite its appearance."

I also verified that my TOTP value rolls over at the same time (within 1 second. There has been 27 leap seconds since 1970 AFAIK) as the value showing on my computers TOTP (KeePassXC).

I will add some comment about this in the code.

[1] https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_xbd_chap04.html#tag_21_04_16